### ISSUE TITLE
A obscure(?) cause to freeze the UI Thread.

### ISSUE THREAD

**Issue by fernandovm** [xamarin/Xamarin.Forms#11519]
After a long, long, long time inspecting my app (including stripping near 95% of they code) I found the cause the problem, but I can't found the explanation, it doesn't make sense for me. Thus, I'm reporting this issue to someone with more knowlodge of xamarin internals to juldge if this is relevant to analize, maybe this can cause other issues hard to find.

### Steps to Reproduce

1. Compile and run the attached sample project [Chronometer.zip](https://github.com/xamarin/xamarin-android/files/4771621/Chronometer.zip).
2. Click in the button 'Start with error'.
3. Wait almost 4 minutes and 30 seconds, satying the screen ON the entire time.

### Expected Behavior

I think could happens two things, or get an error in the method invocation (with an unacceptable parameter value), or all must work indefinitely.

### Actual Behavior

The chronometer follow showing the time progression, but ever near 4.5 min the main thread freezes, without any correlated apparent action.

### Important notes

The sample project is a simple app with two pages, the first is the main page with a custom view (where is the root cause of problem) and two buttons to open the second page with a chronometer.

1. The specific cause of the error is pass in `sweepAngle` parameter of the method `Canvas.DrawArc` the value `NaN`. This code is in the `ProgressRenderer` class.

1. In my tests, any other value avoid the problem, even `Inf`.

1. If I comment the line `this.lblDuration.Text = this.stopwatch.Elapsed.ToString(@"hh\:mm\:ss");` also avoid the problem.

1. I inspected the app in runtime and I can say that none of my app methods was blocked, all they finished normally. I can't see any deadlock.

1. The custom view (`Progress`) is on main page, it is not on the chronometer page that is pushed on top of main page. 

1. The method `OnDraw` of  `ProgressRenderer` class is not invoked continously on each timer interaction but just one time (actually, just in this sample projec it is invoked two times, because the button 'Start with error').

1. I can't see as these apparently uncorrelated events (draw on canvas, in the past, on an hidden page, and a periodic text label update) can together cause the freezing of UI Thread, but they are causing (until where I could see).

1. The background threads reamin working.

1. I'm a bit confuse, the documentation seems say that we need use `Device.BeginInvokeOnMainThread` to change UI elements inside the `Device.StartTimer` callback, but I see tha it already run on UI Thread. The documentation is outdated?

1. At last, after another long random delay (some minutes) after the freeze, I can see the UI Thread go back, moments before an app crash and/or ANR. 

I think that's it, if I remember something else I add in the comments.

### Version Information

Microsoft Visual Studio Enterprise 2019
Version 16.6.1
VisualStudio.16.Release/16.6.1+30128.74
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Enterprise

Xamarin   16.6.000.1061 (d16-6@8e68aa8)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.6.0.318 (remotes/origin/d16-6@66afc5421)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.6.40 (1f1466f)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   10.3.1.4 (d16-6/3a10de9)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: 165f4b0
    Java.Interop: xamarin/java.interop/d16-6@2cab35c
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.31.1@49232bc
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-6@bfb66f3

Xamarin.iOS and Xamarin.Mac SDK   13.18.2.1 (29c4ea7)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by fernandovm** [comment]
Dear @grendello, thank you for attention. Just a note, the issue seems to be device dependent, then, it is important to try testing it on a Samsung Galaxy S6 Edge. Of course, we can't say at this time that the problem is exclusive with the S6 device, only after a precise diagnostic will it be possible to see if it is important to avoid the issue on other device classes.

**Comment by PureWeen** [comment]
@fernandovm if you pause your application while it's blocked and inspect the threads window what's the stack trace of the main thread?

Can you try Android Studio's profiler?

https://developer.android.com/studio/profile/cpu-profiler

See if you can get any insight there to see if the main thread is running/asleep etc.. when the app is frozen?

> the problem is exclusive with the S6 device,

Have you tested on a few different S6 devices? or just one?

**Comment by fernandovm** [comment]
Hi @PureWeen. 

> @fernandovm if you pause your application while it's blocked and inspect the threads window what's the stack trace of the main thread?

Unfortunately this no help, we see just "&lt;not available&gt;".

> Can you try Android Studio's profiler?

Yes, of course. I recorded the profiler session, you can see it here ([cpu-profiler.zip](https://github.com/xamarin/Xamarin.Forms/files/5038639/cpu-profiler.zip)) and the profiler project is here ([com.companyname.chronometer.zip](https://github.com/xamarin/Xamarin.Forms/files/5038644/com.companyname.chronometer.zip)).

> Have you tested on a few different S6 devices? or just one?

I have made tests with two different Samsung S6 Edge devices, see:
![image](https://user-images.githubusercontent.com/441113/89594943-9bf4ec80-d829-11ea-8cab-38422854d80d.png)

**Comment by PureWeen** [comment]
On the last part of your video it switches to the callstack on the threads so we can see more about that nativePollOnce method?

**Comment by fernandovm** [comment]
No @PureWeen, it isn't an attempt to show any specific point, I was searching for anything, but I did not find anything in particular.

Well, if this helps, I can give you a remote access to my device (or my computer) for you to try debug/analyze the issue. Or, if there is any more that I can do, just tell me. I'm very curious about this "obscure" case. :))

**Comment by PureWeen** [comment]
Sorry @fernandovm I phrased that poorly 

Can we see more of the stack trace here?
Can you make more of that visible?

![image](https://user-images.githubusercontent.com/5375137/89688111-bfb64200-d8be-11ea-8d28-8728695e5601.png)


When it freezes is nativePollOnce where you see the time being spent?

**Comment by fernandovm** [comment]
> Sorry @fernandovm I phrased that poorly

Ok, no problem @PureWeen. 

> Can we see more of the stack trace here?
> Can you make more of that visible?

Yes, of course, I will send some more screenshots.

> When it freezes is nativePollOnce where you see the time being spent?

So, I think so (it seems to me just waits).....

![image](https://user-images.githubusercontent.com/441113/89695142-5861b780-d8e9-11ea-9321-d015c93c152c.png)

.... but some things are a bit confusing. After the freeze time we can see a hard single-core cpu usage:

![image](https://user-images.githubusercontent.com/441113/89695555-cc508f80-d8ea-11ea-8a8b-cf6cf95a249f.png)

But the profiler seems not show details about this behavior:

![image](https://user-images.githubusercontent.com/441113/89696716-85b16400-d8ef-11ea-862c-de45e4d89e1b.png)

![image](https://user-images.githubusercontent.com/441113/89696856-0d976e00-d8f0-11ea-90c4-d85dc28a3683.png)

Well, here are the profiler session traces ([cpu-profilers.zip](https://github.com/xamarin/Xamarin.Forms/files/5044124/cpu-profilers.zip), maybe you can see what I didn't see. :))

**Comment by PureWeen** [comment]
@grendello Does any of the profile info that @fernandovm has posted here ring any bells for you?

@fernandovm if you use VS 2017 for deploying your app do you see the same freezes?

**Comment by fernandovm** [comment]
@PureWeen, I don't have VS 2017 installed but can I install it for testing, I must use a specific version or just be the latest version of VS 2017?

**Comment by PureWeen** [comment]
> @PureWeen, I don't have VS 2017 installed but can I install it for testing, I must use a specific version or just be the latest version of VS 2017?

The latest is fine. I'm just curious if a different version of xa gives a different result

**Comment by fernandovm** [comment]
Hi @PureWeen, I made the test and yes, the app freezes also.

![image](https://user-images.githubusercontent.com/441113/89732905-36228380-da28-11ea-9488-80209830b9b5.png)

**Comment by grendello** [comment]
> @grendello Does any of the profile info that @fernandovm has posted here ring any bells for you?
> 

The "unaccounted" spike looks interesting - it might be something that happens in native code, possibly in the Mono runtime (since we have no trace info for it - Android Studio would have it if it was in ART), possibly a large GC collection/related bug? 

Another thing that stands out is the spike in the Render Thread - this has a potential of happening in the rendering daemon, but can be caused by something in the layout (or one of the controls used). A rendering loop between layouts that's either broken forcibly after certain time or is just very long and heavy? An OpenGL shader compilation that's unusually long? 

The last observation is the "Last native thread activity" screenshot - it's interesting that it blocks after posting some message to the Android event queue. It would be interesting to know what the message is and what effects it can have on the process.

**Comment by fernandovm** [comment]
Yeah... At the moment I'm out of ideas, I don't see how to go deeper. ;-((  Anyway, I am available to carry out any new tests that are necessary, or even, to grant remote access to my devices. ;-))

**Comment by AntRemo** [comment]
Hi @fernandovm 

I know this is old... but was curious if you were able to resolve it or had any other insights about 

> 9. I'm a bit confuse, the documentation seems say that we need use Device.BeginInvokeOnMainThread to change UI elements inside **the Device.StartTimer callback, but I see tha it already run on UI Thread. The documentation is outdated?**

Many Thanks 🙏

**Comment by fernandovm** [comment]
Hi @AntRemo.. This point was just a little curiosity, but I can't belive that was direct related with the issue.
